service-deployment/bootstrap/prometheus/prometheus-configmap.yaml.template

```diff
@@ -23,18 +23,22 @@ data:
   prometheus.yml: |-
     # Scrape config for cluster components.
     scrape_configs: 
-    - job_name: "node_exporter"
-      scrape_interval: "15s"
-      static_configs:
-      - targets: 
-        {% for host in machinelist %}
-        - {{ machinelist[host]['ip']}}:{{ clusterinfo['prometheusinfo']['node_exporter_port'] }}
-        {% endfor %}
-    - job_name: "cadvisor"
-      scrape_interval: "15s"
-      static_configs:
-      - targets: 
-        {% for host in machinelist %}
-        - {{ machinelist[host]['ip']}}:{{ clusterinfo['prometheusinfo']['cadvisor_port'] }}
-        {% endfor %}
+    - job_name: 'node_exporter'
+      kubernetes_sd_configs:
+      - api_server: '{{ clusterinfo['webportalinfo']['k8s_api_server_uri'] }}'
+        role: node
+      relabel_configs:
+      - source_labels: [__address__]
+        regex: '(.*):10250'
```

Comment:
>10250 [](start = 21, length = 5)

what does 10250 mean? can you explain the purpose of relabel_configs?

Reply:
prometheus will defaultly get each node __address__, it contains 10250 port, then replace it with exporter port to fetch metrics.
![image](https://user-images.githubusercontent.com/5576848/37749052-1cbd25f0-2dc1-11e8-9b66-6c634913997a.png)

Also ref: https://github.com/prometheus/docs/issues/836

Reply:
this means port 10250 is reserved? we should state this in prom deployment service to avoid other service accidently uses it.

---
In reply to: [176298014](https://github.com/Microsoft/pai/pull/395#discussion_r176298014) [](ancestors = 176298014)

Reply:
i check 10250 was reserved. this will add a comment at clusterconfig-example.json

Reply:
in here: [https://github.com/Microsoft/pai/blob/master/service-deployment/clusterconfig-example.yaml#L121](https://github.com/Microsoft/pai/blob/master/service-deployment/clusterconfig-example.yaml#L121), you should add a field "prometheus_reserved_port" and set it to 10250, and add a comment saying don't change it. This way, a tool can parse all the sections and detect whether there exist port confliction.

---
In reply to: [176301229](https://github.com/Microsoft/pai/pull/395#discussion_r176301229) [](ancestors = 176301229)

Reply:
10250 is the default Kubelet port. For us, I think we can just ignore any port part in __address__ and add with the exporters port specified in our config file.

---
In reply to: [176300755](https://github.com/Microsoft/pai/pull/395#discussion_r176300755) [](ancestors = 176300755,176298014)

Reply:
yes, prometheus call k8s api to get this port and output the __address__ tag. currently, the 10250 has been remove and use a regex to cover any port